### ISSUE TITLE
[Question] How to get same logits as HuggingFace gpt2?

### ISSUE THREAD

**Issue by kmrasmussen** [neelnanda-io/TransformerLens#223]
### Question

If have a sentence such as 'hi how are you', I want to use get the final output logits from a model (ideally Pythia, put I also have problems with gpt2). I do not yet know all the in and outs of both transformers and transformer_lens but it seems there is some unintuitive mismatch between the logits I get from gpt2 on standard transformers gpt2 and hookedtransformer gpt2. I have a notebook https://colab.research.google.com/drive/1orG4B2An6p7l3kc92RXWV1FOodIUlKwl?usp=sharing

```python
import torch
from transformers import GPT2Tokenizer, GPT2Model, GPT2LMHeadModel
from transformer_lens import HookedTransformer

tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2LMHeadModel.from_pretrained('gpt2', return_dict=True)
model_lens = HookedTransformer.from_pretrained("gpt2")

sentence = 'hi how are you'
input_ids = tokenizer.encode(sentence, return_tensors='pt')

```
```
> model_lens(input_ids, return_type='logits')
tensor([[[ 5.9978,  6.2685,  3.3547,  ..., -1.4218, -0.4220,  5.4836],
         [ 7.1920,  6.1621,  0.3314,  ..., -0.5739,  0.5229,  3.9193],
         [ 5.5006,  4.7137,  0.9853,  ...,  1.6279, -1.1324,  3.6218],
         [10.5213,  8.5669,  2.3738,  ...,  0.3045, -0.8924,  4.7973]]],
       grad_fn=<AddBackward0>)
```
which is same results as ``model_lens(sentence, prepend_bos=False)```, and

```
> model_lens(sentence, prepend_bos=True)
tensor([[[ 7.5261, 11.1214,  7.8919,  ..., -3.1299, -3.3873,  8.5934],
         [ 9.7903,  6.6537,  3.4986,  ..., -2.3309,  0.4245,  7.2998],
         [ 7.4576,  6.9958,  3.2204,  ..., -1.6337, -1.5720,  7.0206],
         [ 4.9628,  3.7675,  2.3309,  ...,  0.6831, -1.6884,  4.5932],
         [10.1804,  6.8850,  4.1714,  ..., -0.3637, -1.8858,  6.7364]]],
       grad_fn=<AddBackward0>)
```

while with the transformer model I get
```
> model(input_ids).logits
tensor([[[ -31.3108,  -31.0401,  -33.9539,  ...,  -38.7304,  -37.7306,
           -31.8250],
         [ -85.6411,  -86.6710,  -92.5017,  ...,  -93.4070,  -92.3102,
           -88.9139],
         [ -85.9755,  -86.7625,  -90.4908,  ...,  -89.8483,  -92.6085,
           -87.8544],
         [ -92.9654,  -94.9198, -101.1129,  ..., -103.1822, -104.3792,
           -98.6894]]], grad_fn=<UnsafeViewBackward0>)
```

Am I missing some piece about how tokenization works in both of these libraries?

**Comment by neelnanda-io** [comment]
The main difference is that TL makes the logits have mean zero, which
essentially adds a constant at each position. You can see that the
difference in logits at each position is identical between the two runs
(though it's different when you prepend Bos, because that's a different
input). This behaviour exists because the log probs are unchanged, as
softmax is invariant on adding a constant, and in my opinion the mean is
distracting and irrelevant information that's better to remove.

You can disable this behaviour by passing center_unembed=False to
from_pretrained or by using from_pretrained_no_processing

**Comment by kmrasmussen** [comment]
Thank you, Neel!
